jmespath-core/src/main/java/io/burt/jmespath/function/JoinFunction.java

```diff
@@ -0,0 +1,57 @@
+package io.burt.jmespath.function;
+
+import java.util.List;
+import java.util.ArrayList;
+
+import io.burt.jmespath.Adapter;
+
+public class JoinFunction extends JmesPathFunction {
+  public JoinFunction() {
+    super(2, 2);
+  }
+
+  @Override
+  protected <T> T internalCall(Adapter<T> adapter, List<ExpressionOrValue<T>> arguments) {
+    T glue = arguments.get(0).value();
+    T components = arguments.get(1).value();
+    if (isStringArray(adapter, components)) {
+      if (adapter.isString(glue)) {
+        List<T> values = adapter.toList(components);
+        if (values.isEmpty()) {
+          return adapter.createString("");
+        } else {
+          StringBuffer buffer = new StringBuffer();
+          String glueString = adapter.toString(glue);
+          for (T value : values) {
+            buffer.append(adapter.toString(value));
+            buffer.append(glueString);
+          }
+          buffer.delete(buffer.length() - glueString.length(), buffer.length());
```

Comment:
Given that values is a `List<T>` this is a bit sad, as the final glue string could cause the entire buffer to be reallocated (unlikely in the general case, but it is deterministic, so for example a list with only one 16-character string would cause allocation of a 34-character string for no real reason). It's extra sad that for a non-jfc adapter, the intermediate resulting string will probably be "unnecessary" as well.